### PR TITLE
helper-cli: CreateCommand improvements

### PR DESCRIPTION
**Pull request by fviernau**
See individual commits.

### CODE REVIEW

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt [hunk 1]
```diff
@@ -84,21 +84,24 @@ internal class CreateCommand : CliktCommand(
...
+
+        scanResults.find { it.provenance is ArtifactProvenance }?.let {
+            createPackageConfiguration(it.provenance)
+        }?.writeToFile()
     }
 
-    private fun Provenance.writePackageConfigurationFile(filename: String) {
-        val packageConfiguration = createPackageConfiguration(this)
+    private fun PackageConfiguration.writeToFile() {
+        val filename = "vcs.yml".takeIf { vcs != null } ?: "source-artifact.yml"
```
Comment (mnonnenmacher) [inline helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:98]: I find `val filename = if (vcs != null) "vcs.yml" else "source-artifact.yml"` easier to read, and it's also shorter.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

**Review by mnonnenmacher — APPROVED** [review]

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt [hunk 2]
```diff
@@ -113,25 +113,17 @@ internal class CreateCommand : CliktCommand(
...
 
     private fun createPackageConfiguration(provenance: Provenance): PackageConfiguration =
-        when (provenance) {
-            is ArtifactProvenance -> {
-                PackageConfiguration(
-                    id = packageId,
-                    sourceArtifactUrl = provenance.sourceArtifact.url
+        PackageConfiguration(
+            id = packageId,
+            sourceArtifactUrl = provenance.let { it as? ArtifactProvenance }?.let {
```
Comment (sschuberth) [inline helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:122]: Simplify to: `(provenance as? ArtifactProvenance)?.sourceArtifact?.url`

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt [hunk 3]
```diff
@@ -113,25 +113,17 @@ internal class CreateCommand : CliktCommand(
...
+                it.sourceArtifact.url
+            },
+            vcs = provenance.let { it as? RepositoryProvenance }?.let {
```
Comment (sschuberth) [inline helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:125]: Simplify to: `(provenance as? RepositoryProvenance)?.let {`

**Comment by fviernau** [comment]
Merged despite the unrelated failed tests.

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt [hunk 4]
```diff
@@ -82,15 +82,16 @@ internal class CreateCommand : CliktCommand(
...
         val scanResultsStorage = FileBasedStorage(LocalFileStorage(scanResultsStorageDir))
-        val scanResults = scanResultsStorage.read(packageId).getOrDefault(emptyList())
-
-        scanResults.find { it.provenance is RepositoryProvenance }?.let {
-            createPackageConfiguration(it.provenance)
-        }?.writeToFile()
+        val scanResults = scanResultsStorage.read(packageId).getOrDefault(emptyList()).run {
+            listOfNotNull(
+                find { it.provenance is RepositoryProvenance },
+                find { it.provenance is ArtifactProvenance }
```
Comment (sschuberth) [inline helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:88]: Can't you simply filter for `KnownProvenance`?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt (hunk 2 above)
Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:122]: It seems while you reviewed I merged already. I've addressed this with https://github.com/oss-review-toolkit/ort/pull/5687.

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt (hunk 3 above)
Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:125]: It seems while you reviewed I merged already. I've addressed this with https://github.com/oss-review-toolkit/ort/pull/5687.

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt (hunk 4 above)
Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:88]: The current logic ensures there is at most one RepositoryProvenance and one ArtifactProvenance.
Filtering for KnownProvenance has different semantics IIUC.

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt (hunk 2 above)
Comment (sschuberth) [reply helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:122]: The code in that PR is still more complicated than my proposal here.

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt (hunk 4 above)
Comment (sschuberth) [reply helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:88]: > The current logic ensures there is at most one RepositoryProvenance and one ArtifactProvenance.

I don't think it does: `find` will get you the *first* of either provenance type, if any, but there could be multiple.

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt (hunk 2 above)
Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:122]: Oh sorry, it was too late yesterday. I'll fix it up.

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt (hunk 4 above)
Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:88]: `find { it.provenance is RepositoryProvenance }` will return the first provenance which is an instance of `RepositoryProvenance` or null. The second find call will do the same for `ArtifactProvenance`. 
So constraints of the built list are:

- contains at most 2 items
- contains at most one instance of `ArtifactProvenance`
- contains at most one instance of `RepositoryProvenance`

If one uses filter by known provenance instead, above contraints will get eliminated.
With above constraints: No file can be written twice as `ArtifactProvenance` goes to `source-artifact.yml` and `RepositoryProvenance` to "vcs.yml".

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt (hunk 2 above)
Comment (fviernau) [reply helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:122]: Hope this is it now: https://github.com/oss-review-toolkit/ort/pull/5689

File: helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt (hunk 4 above)
Comment (sschuberth) [reply helper-cli/src/main/kotlin/commands/packageconfig/CreateCommand.kt:88]: > The current logic ensures there is at most one RepositoryProvenance and one ArtifactProvenance.

Ah, I misunderstood what "there is" refers to: You refer to the output list, whereas I thought you'd be referring to the input list.

All good now!

**Pull request merged**